flashinfer/gemm.py

```diff
@@ -349,6 +358,216 @@ def get_gemm_sm100_module():
+            scale_granularity_n=scale_n,
+            scale_granularity_k=scale_k,
+        )
+        write_if_different(dest_path, source)
+
+    # Generate group gemm kernel instantiations
+    prefix = "group_gemm_fp8_groupwise"
+    with open(jit_env.FLASHINFER_CSRC_DIR / f"{prefix}_sm120_kernel_inst.jinja") as f:
+        kernel_inst_templ = jinja2.Template(f.read())
+
+    # Use the same mma_sm_list and scale_granularity_configs as regular gemm
+    for dtype_in, dtype_out, scale_major_k, mma_sm, (scale_m, scale_n, scale_k) in product(
+        dtype_in_list, dtype_out_list, scale_major_k_list, mma_sm_list, scale_granularity_configs
+    ):
+        name_dtype_in = filename_safe_dtype_map[dtype_in]
+        name_dtype_out = filename_safe_dtype_map[dtype_out]
+        dest_path = (
+            gen_directory
+            / f"{prefix}_{name_dtype_in}_{name_dtype_out}_major{scale_major_k}_mma{mma_sm}_m{scale_m}_n{scale_n}_k{scale_k}_sm120.cu"
+        )
+        source_paths.append(dest_path)
+        source = kernel_inst_templ.render(
+            dtype_in=dtype_cutlass_map[dtype_in],
+            dtype_out=dtype_cutlass_map[dtype_out],
+            scale_major_k=scale_major_k,
+            mma_sm=mma_sm,
+            scale_granularity_m=scale_m,
+            scale_granularity_n=scale_n,
+            scale_granularity_k=scale_k,
+        )
+        write_if_different(dest_path, source)
+
+    # Copy source files
+    for filename in [
+        "gemm_groupwise_sm120.cu",
+        "group_gemm_fp8_groupwise_sm120.cu",
+        "gemm_sm120_pybind.cu",
+        "group_gemm_sm120_pybind.cu",
+    ]:
+        src_path = jit_env.FLASHINFER_CSRC_DIR / filename
+        dest_path = gen_directory / filename
+        source_paths.append(dest_path)
+        with open(src_path, "r") as f:
+            source = f.read()
+        write_if_different(dest_path, source)
+
+    print(f"Debugging SM120/SM121 NVCC flags: {sm121a_nvcc_flags}")
+
+    return gen_jit_spec(
+        "gemm_sm120",
+        source_paths,
+        extra_cuda_cflags=sm121a_nvcc_flags, # TODO (yongwww): fix
+    )
+
+
+@functools.cache
+def get_gemm_sm120_module():
+    module = gen_gemm_sm120_module().build_and_load()
+    return module
+
+
+@functools.cache
+def get_gemm_sm120_module_cutlass_fp8():
+    """Get CUTLASS FP8 runner for SM120/SM121 using the groupwise scaling kernel."""
+    module = get_gemm_sm120_module()
+    
+    def cutlass_fp8_gemm_runner():
+        class CutlassFp8GemmRunner(TunableRunner):
+            def get_valid_tactics(
+                self,
+                inputs: List[torch.Tensor],
+                profile: OptimizationProfile,
+            ) -> List[int]:
+                # For now, return a single default tactic
+                return [-1]
+            
+            def forward(
+                self,
+                inputs: List[torch.Tensor],
+                tactic: int = -1,
+                do_preparation: bool = False,
```

Comment:
seems like this parameter is unused?

Reply:
Good catch. They’re part of TunableRunner, keeping them for consistency with the others.